server/sonar-server/src/main/java/org/sonar/server/settings/ws/SettingsLoader.java

```diff
@@ -0,0 +1,109 @@
+  }
+
+  public List<Setting> loadGlobalSettings(DbSession dbSession, Set<String> keys) {
+    List<PropertyDto> properties = dbClient.propertiesDao().selectGlobalPropertiesByKeys(dbSession, keys);
+    List<PropertyDto> propertySets = dbClient.propertiesDao().selectGlobalPropertiesByKeys(dbSession, getPropertySetKeys(properties));
+    return properties.stream()
+      .map(property -> new Setting(property, getPropertySets(property.getKey(), propertySets, null), definitions.get(property.getKey())))
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * Return list of settings by component uuid, sorted from project to lowest module
+   */
+  public Multimap<String, Setting> loadComponentSettings(DbSession dbSession, Set<String> keys, ComponentDto component) {
+    List<String> componentUuids = DOT_SPLITTER.splitToList(component.moduleUuidPath());
+    List<ComponentDto> componentDtos = dbClient.componentDao().selectByUuids(dbSession, componentUuids);
+    Set<Long> componentIds = componentDtos.stream().map(ComponentDto::getId).collect(Collectors.toSet());
+    Map<Long, String> uuidsById = componentDtos.stream().collect(Collectors.toMap(ComponentDto::getId, ComponentDto::uuid));
+    List<PropertyDto> properties = dbClient.propertiesDao().selectPropertiesByKeysAndComponentIds(dbSession, keys, componentIds);
+    List<PropertyDto> propertySets = dbClient.propertiesDao().selectPropertiesByKeysAndComponentIds(dbSession, getPropertySetKeys(properties), componentIds);
+
+    Multimap<String, Setting> settingsByUuid = TreeMultimap.create(Ordering.explicit(componentUuids), Ordering.arbitrary());
+    for (PropertyDto propertyDto : properties) {
+      Long componentId = propertyDto.getResourceId();
+      String componentUuid = uuidsById.get(componentId);
+      String propertyKey = propertyDto.getKey();
+      settingsByUuid.put(componentUuid,
+        new Setting(propertyDto, getPropertySets(propertyKey, propertySets, componentId), definitions.get(propertyKey)));
+    }
+    return settingsByUuid;
+  }
+
+  private Set<String> getPropertySetKeys(List<PropertyDto> properties) {
+    Set<String> propertySetKeys = new HashSet<>();
+    properties.stream()
+      .filter(propertyDto -> definitions.get(propertyDto.getKey()) != null)
+      .filter(propertyDto -> definitions.get(propertyDto.getKey()).type().equals(PROPERTY_SET))
+      .forEach(propertyDto -> definitions.get(propertyDto.getKey()).fields()
+        .forEach(field -> COMMA_SPLITTER.splitToList(propertyDto.getValue())
+          .forEach(setId -> propertySetKeys.add(generatePropertySetKey(propertyDto.getKey(), setId, field.key())))));
+    return propertySetKeys;
+  }
+
+  private static String generatePropertySetKey(String propertyKey, String id, String fieldKey) {
+    return propertyKey + "." + id + "." + fieldKey;
+  }
+
+  private static List<PropertyDto> getPropertySets(String propertyKey, List<PropertyDto> propertySets, @Nullable Long componentId) {
```

Comment:
`propertyKey` or `propertyDto` should be renamed. One is from the root property, the other is from a property set, which is a property